nse_main.lua

```diff
@@ -661,6 +661,75 @@ do
   Script.__index = Script;
 end
 
+local function generate_args(tbl)
+  local o={}
+  local name=""
+  for _, val in pairs(tbl) do
+    if type(val) == "string" then
+      name=val
+    end
+    if type(val) == "table" then
+      table.insert(o,name.."."..val[1].."="..val[2])
+    end
+  end
+  return o
+end
+
+local function get_script_name(tbl)
+  local o={}
+  for _, val in pairs(tbl) do
+    if type(val) == "string" then
+      table.insert(o,val)
+    end
+  end
+  return o
+end
+
+local function make_parser()
```

Comment:
Instead of re-implementing script argument parsing, perhaps we can reuse the arg parser [1]. i.e. move the script argument parser into another function which is used in called in both locations.

After making this change [2], you will want to change `get_chosen_scripts` [3] to parse the new `--script` argument which is not split into a table for us by C++ (instead, it's the same string passed on the command line). The rule string will include normal rules but script names (not categories or script expressions) can have optional arguments. (Looking at the code more, this could be a very complicated modification... it will require more discussion.)

[1] https://github.com/tremblerz/nmap/blob/7c637f5cb35e553306dc38a73b734e47dfe35a35/nse_main.lua#L1259-L1274
[2] https://github.com/nmap/nmap/pull/415/files#r66714914
[3] https://github.com/tremblerz/nmap/blob/7c637f5cb35e553306dc38a73b734e47dfe35a35/nse_main.lua#L750